src/snapred/backend/recipe/FetchGroceriesRecipe.py

```diff
@@ -17,52 +18,113 @@
 
-    def _runKey(self, runConfig: RunConfig) -> Tuple[str, str, bool]:
-        return (runConfig.runNumber, runConfig.IPTS, runConfig.isLite)
+    def _key(self, item: GroceryListItem) -> Tuple[str, bool]:
+        if item.workspaceType == "nexus":
+            return (item.runNumber, item.useLiteMode)
+        elif item.workspaceType == "grouping":
+            return (item.groupingScheme, item.useLiteMode)
 
-    def _createFilenameFromRunConfig(self, runConfig: RunConfig) -> str:
-        instr = "nexus.lite" if runConfig.isLite else "nexus.native"
+    ## FILENAME METHODS
+
+    def _createNexusFilename(self, item: GroceryListItem) -> str:
+        instr = "nexus.lite" if item.useLiteMode else "nexus.native"
         pre = instr + ".prefix"
         ext = instr + ".extension"
-        return runConfig.IPTS + Config[pre] + str(runConfig.runNumber) + Config[ext]
+        return item.IPTS + Config[pre] + str(item.runNumber) + Config[ext]
+
+    def _createGroupingFilename(self, item: GroceryListItem):
+        if item.groupingScheme == "Lite":
+            return str(Config["instrument.lite.map.file"])
+        instr = "lite" if item.useLiteMode else "native"
+        home = "instrument.calibration.powder.grouping.home"
+        pre = "grouping.filename.prefix"
+        ext = "grouping.filename." + instr + ".extension"
+        return Config[home] + "/" + Config[pre] + item.groupingScheme + Config[ext]
+
+    ## WORKSPACE NAME METHODS
 
-    def _createNexusWorkspaceNameBuilder(self, runConfig: RunConfig) -> NameBuilder:
-        runNameBuilder = wng.run().runNumber(runConfig.runNumber)
-        if runConfig.isLite:
+    def _createNexusWorkspaceNameBuilder(self, item: GroceryListItem) -> NameBuilder:
+        runNameBuilder = wng.run().runNumber(item.runNumber)
+        if item.useLiteMode:
             runNameBuilder.lite(wng.Lite.TRUE)
         return runNameBuilder
 
-    def _createNexusWorkspaceName(self, runConfig: RunConfig) -> str:
-        return self._createNexusWorkspaceNameBuilder(runConfig).build()
+    def _createNexusWorkspaceName(self, item: GroceryListItem) -> str:
+        return self._createNexusWorkspaceNameBuilder(item).build()
+
+    def _createRawNexusWorkspaceName(self, item: GroceryListItem) -> str:
+        return self._createNexusWorkspaceNameBuilder(item).auxilary("Raw").build()
+
+    def _createCopyNexusWorkspaceName(self, item: GroceryListItem, numCopies: int) -> str:
+        return self._createNexusWorkspaceNameBuilder(item).auxilary(f"Copy{numCopies}").build()
 
-    def _createRawNexusWorkspaceName(self, runConfig: RunConfig) -> str:
-        return self._createNexusWorkspaceNameBuilder(runConfig).auxilary("Raw").build()
+    def _createGroupingWorkspaceName(self, item: GroceryListItem):
+        if item.groupingScheme == "Lite":
+            return "lite_grouping_map"
+        instr = "lite" if item.useLiteMode else "native"
+        return Config["grouping.workspacename." + instr] + item.groupingScheme
```

Comment:
Should this sort of logic live in the WorkspaceNameGenerator?

Reply:
a workspace name generator for it would be good, the nexus data already has name generators